### DISCUSSION TITLE
How to handle UUID in SQLModel for MSSQL ? (Conversion failed)

### DISCUSSION THREAD

**Discussion by Ringostar1982** [fastapi/sqlmodel#878] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
Model(shortened):
from pydantic import UUID4

class Customer_RoleBase(SQLModel):
    roles_id: Optional[UUID4] = Field(primary_key=True, nullable=False)

Function to update DB(shortened):
def update_customer_role(role: Customer_Role) -> Customer_Role:
    with Session(createDBEngine()) as session:        
        print(role.roles_id)
        print(type(role.roles_id))
       
        target = session.exec(select(Customer_Role).where(Customer_Role.roles_id == role.roles_id)).one()
        
        if not target:
            raise HTTPException(status_code=404, detail="Customer-Role not found")
```

### Description

- MSSQL Server
- DB Driver pymssql: 2.2.11
- SQLModel: 0.0.11
- pydantic: 1.10.13
- fastapi: 0.104.1
- SQLAlchemy: 1.4.50

I just try to get the matching role by its roles_id (UNIQUEIDENTIFIER) but it end´s in an error:

### Console Output: (see the print commands are first)

120ba15d-61ea-40fa-8a39-dd0145d7ec7d
<class 'uuid.UUID'>

**sqlalchemy.exc.OperationalError: (pymssql._pymssql.OperationalError) (8169, b'Conversion failed when converting from a character string to uniqueidentifier.DB-Lib error message 20018**

[SQL: SELECT customer_roles.roles_id, customer_roles.customer_id, customer_roles.system_id, customer_roles.salesorg_id, customer_roles.customer_role, customer_roles.role_number, customer_roles.full_name, customer_roles.first_name, customer_roles.last_name, customer_roles.partner_desc, customer_roles.default_partner, customer_roles.street, customer_roles.zip, customer_roles.city, customer_roles.country_code, customer_roles.email, customer_roles.phone, customer_roles.currency_code, customer_roles.payment_code, customer_roles.inco_terms_1, customer_roles.inco_terms_2, customer_roles.dist_channel, customer_roles.division
FROM customer_roles
WHERE customer_roles.roles_id = %(roles_id_1)s]
[parameters: {'roles_id_1': '120ba15d61ea40fa8a39dd0145d7ec7d'}]

The Column roles_id is of Type UUID4 and in the print command we can see its a real UUID and type of uuid.UUID but something after will convert it to something like a string without the hyphens.

If i copy the SQL Statement to Azure Data Studio and run the query, it results in same error.
If i include the hyphens to the query it works.

I was searching and testing until now alot but could not find a correct solution.

### Operating System

Windows

### SQLModel Version

0.0.11

### Python Version

3.11.0

**Comment by Ringostar1982** [comment]
update:

If I CAST the roles_id to UNIQUEIDENTIFIER  the Where Clause will be ok, but later on the UPDATE Command will fail!

```
from sqlalchemy.dialects.mssql import UNIQUEIDENTIFIER

target = session.exec(select(Customer_Role).where(
            Customer_Role.roles_id == sql.func.CAST(
                role.roles_id, UNIQUEIDENTIFIER
                )
            )).one()
```


The next steps in updating the target will fail:
```
new_data = role.dict(exclude_unset=True)
        for key, value in new_data.items():
            setattr(target, key, value)
        Customer_Role.from_orm(target)  # above SQLModel v2: Customer.model_validate(target)
        # print(target.json(by_alias=True))
        session.add(target)
        session.commit()  # Commit all the changes tracked in the session. This commits everything in one single batch.
        session.refresh(target)  # update the object!
        return target
```

Error Message:

`[SQL: UPDATE customer_roles SET street=%(street)s WHERE customer_roles.roles_id = %(customer_roles_roles_id)s]      
[parameters: {'street': 'Straße 13', 'customer_roles_roles_id': '120ba15d61ea40fa8a39dd0145d7ec7d'}]`

**Reply by joachimhuet** [reply]
I was facing the same problem and could not find a solution so temporarily I did the following:
```
id: str = Field(
        primary_key=True,
        default_factory=lambda: str(uuid4())[:36],
        max_length=36,
        unique=True,
    )
```

**Reply by Ringostar1982** [reply]
Hello Joachim, in my case the Database handles the autocreation of Field roles_id (uniqueidentifier) by its default SQL Server Function => `newid()`
and therefore i declared the roles_id as `Optional[ ]`

I could test if following will work:

`roles_id: Optional[str] = Field(primary_key=True, default_factory=lambda: str(uuid4())[:36], max_length=36, nullable=False)`
`# unique=True `

Also i have to change the DB- Column to String `(varchar(50)` and its default-Function now should create an uniqueidentifier and convert it to String. Maybe something like that:
`CONVERT(CHAR(255), newid()) AS 'char'` instead of` newid()`

**Reply by joachimhuet** [reply]
Well in fact the problem is just that the UUID if not converted to a string won't contain the hyphens so that's why I used this 'hack'. But you shouldn't need to change your whole database for that.
**Edit:**
You can check that sqlalchemy thread for more information on this topic: https://github.com/sqlalchemy/sqlalchemy/discussions/10336
According to this thread you indeed need to change the column type 😞

**Reply by Ringostar1982** [reply]
That seems to be working... for the moment by updating a Row ... creating new Rows needs to be tested.
`import uuid`
` roles_id: Optional[str] = Field(primary_key=True, default_factory=lambda: str(uuid.uuid4())[:36], max_length=36, nullable=False, unique=True)`

I changed database Column to varchar(100) and default is now `(CONVERT([varchar](100),newid()))`

**Reply by joachimhuet** [reply]
If the database is responsible for creating that id you don't need the default_factory. In my case it was my code taking care of generating that id. You can use the sqlalchemy `server_default` parameter instead but it's not necessary if you created your database from sql code directly and not from your python. Because you will already have it defined there.

**Reply by Ringostar1982** [reply]
So in my case it should be enaught to declare ?
`roles_id: Optional[str] = Field(primary_key=True, nullable=False)`

**Reply by joachimhuet** [reply]
Yes I would say so, and to be precise you could add your `max_length = 100` that is the size of your column. But as it's an id you won't edit it so it will always be given by the database and should never be more than 100 anyway.
The more precise you are in your sqlmodel the easier it would be for someone else to understand your database without needing to look at your sql so I would still encourage do match your database description in sqlmodel.

For example if one day you want to spawn a test database you could just use your sqlmodel to create it. And having the `server_default` defined will add it directly on create table statement.
I don't know if my point is clear, but as an answer yes you could do that but you probably should not 🤣

**Reply by Ringostar1982** [reply]
In your understanding what would you recommend to be best in my case ?
`roles_id: Optional[str] = Field(primary_key=True, nullable=False, server_default, max_length = 100)`

**Reply by joachimhuet** [reply]
How did you make your id to be automatically generated in sql ? Using `NEWID()` on table create statement ? If so I would write something like that (nullable and unique are already suggested by `primary_key=True` [thread](https://stackoverflow.com/questions/58139212/django-does-primary-key-true-also-mean-unique)):
`roles_id: Optional[str] = Field(primary_key=True, sa_column_kwargs={"server_default": "NEWID()"}, max_length=100)`

**Comment by Ringostar1982** [comment]
The DB-Column´s default value = `(CONVERT([varchar](100),newid()))`

`roles_id: Optional[str] = Field(primary_key=True, sa_column_kwargs={"server_default": "(CONVERT([varchar](100),newid()))"}, max_length=100)` 

The server_default property is just for information?

**Reply by joachimhuet** [reply]
> The server_default property is just for information?
Yes for your typical case But definitely **NO** because it is used by sqlalchemy and added as metadata to your table.

See this sqlalchemy example without server_default:
```
from sqlalchemy import Column, String
from sqlalchemy.schema import CreateTable
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.dialects import mssql

Base = declarative_base()

class Roles(Base):
    __tablename__ = "roles"

    id = Column(String(100), primary_key=True)
    name = Column(String(100), nullable=False)
    description = Column(String(100))
    
print(CreateTable(Roles.__table__).compile(dialect=mssql.dialect()))
"""Produces the following sql code:
CREATE TABLE roles (
	id VARCHAR(100) NOT NULL, 
	name VARCHAR(100) NOT NULL, 
	description VARCHAR(100) NULL, 
	PRIMARY KEY (id)
)
"""
```

But now if you add server_default parameter:
```
class Roles(Base):
    __tablename__ = "roles"

    id = Column(String(100), primary_key=True, server_default="(CONVERT([varchar](100),newid()))")
    name = Column(String(100), nullable=False)
    description = Column(String(100))

print(CreateTable(Roles.__table__).compile(dialect=mssql.dialect()))
"""Now produces:
CREATE TABLE roles (
	id VARCHAR(100) NOT NULL DEFAULT '(CONVERT([varchar](100),newid()))', 
	name VARCHAR(100) NOT NULL, 
	description VARCHAR(100) NULL, 
	PRIMARY KEY (id)
)
"""
```

So it means you can create your table directly from python and have exactly the same as your current table.

Oh and a small detail I forgot to mention, you need to put `default=None` on your field, because otherwise When you'll try to instantiate a Role object it will ask you for the `roles_id` that you won't have at this point.

Is everything clear now ?

**Reply by Ringostar1982** [reply]
I think you forgot to delete the server_default in your upper example without server_default.

So in my case i go on with(i wont do creating tables by my code for now):
`roles_id: Optional[str] = Field(primary_key=True, default=None, sa_column_kwargs={"server_default": "(CONVERT([varchar](100),newid()))"}, max_length=100)`

Editing Data was succefully. Next is to test if creating new Rows also will work.

**Reply by joachimhuet** [reply]
You are right sorry about that 💤 
Great tell me if you need help for anything else !

**Reply by Ringostar1982** [reply]
Thanks, I will reply as soon as possible with results ..

**Reply by Ringostar1982** [reply, marked as answer]
Actually the workaround solution is working for updating and creating new rows in DB-table.
@joachimhuet thanks for support!

```
class Customer_RoleBase(SQLModel):
    roles_id: Optional[str] = Field(primary_key=True, index=True, default=None, sa_column_kwargs={"server_default": "(CONVERT([varchar](100),newid()))"}, max_length=100)
    ...
```

**Reply by urosjarc** [reply]
@joachimhuet & @Ringostar1982 Why we can't use roles_id: UUID ? What is the reason to have column as str type?
I have the same issue but I don't like having str type as column.

**Reply by redb0** [reply]
@urosjarc How about using `TypeDecorator` (https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator)? This will convert the `UUID` type to a database-only string. On the Python side, you will still have a `UUID`

**Comment by michelmetran** [comment]
Thank you very much!
I was looking for a solution... I had already tried dozens of things.
